adafruit_bluefruitspi.py

```diff
@@ -0,0 +1,220 @@
+    - ``ArgType.STRING``
+    - ``ArgType.BYTEARRAY``
+    - ``ArgType.INT32``
+    - ``ArgType.UINT32``
+    - ``ArgType.INT16``
+    - ``ArgType.UINT16``
+    - ``ArgType.INT8``
+    - ``ArgType.UINT8``
+    """
+    STRING    = const(0x0100) # String data type
+    BYTEARRAY = const(0x0200) # Byte array data type
+    INT32     = const(0x0300) # Signed 32-bit integer data type
+    UINT32    = const(0x0400) # Unsigned 32-bit integer data type
+    INT16     = const(0x0500) # Signed 16-bit integer data type
+    UINT16    = const(0x0600) # Unsigned 16-bit integer data type
+    INT8      = const(0x0700) # Signed 8-bit integer data type
+    UINT8     = const(0x0800) # Unsigned 8-bit integer data type
+
+
+class ErrorCode:  #pylint: disable=too-few-public-methods,bad-whitespace
+    """An enum-like class representing possible error codes.
+    Possible values are
+    - ``ErrorCode.``
+    """
+    INVALIDMSGTYPE = const(0x8021) # SDEP: Unexpected SDEP MsgType
+    INVALIDCMDID   = const(0x8022) # SDEP: Unknown command ID
+    INVALIDPAYLOAD = const(0x8023) # SDEP: Payload problem
+    INVALIDLEN     = const(0x8024) # SDEP: Indicated len too large
+    INVALIDINPUT   = const(0x8060) # AT: Invalid data
+    UNKNOWNCMD     = const(0x8061) # AT: Unknown command name
+    INVALIDPARAM   = const(0x8062) # AT: Invalid param value
+    UNSUPPORTED    = const(0x8063) # AT: Unsupported command
+
+
+class BluefruitSPI:
+    """Helper for the Bluefruit LE SPI Friend"""
+
+    def __init__(self, spi, cs, irq, debug=False):
+        self._cs = cs
+        self._irq = irq
+        self._spi = spi
+        self._ble = SPIDevice(self._spi, self._cs)
+        self._buf_tx = bytearray(20)
+        self._buf_rx = bytearray(20)
+        self._debug = debug
+
+        # CS is an active low output, so set pullup
+        self._cs.switch_to_output(value=True,
+                                  drive_mode=digitalio.DriveMode.PUSH_PULL)
+
+        # irq line is active high input, so set a pulldown as a precaution
+        self._irq.switch_to_input(pull=digitalio.Pull.DOWN)
+
+        # Check out the SPI bus
+        while not self._spi.try_lock():
+            pass
+
+        # Configure SPI for 4MHz
+        self._spi.configure(baudrate=4000000, phase=0, polarity=0)
+
+        self._spi.unlock()
+
+    def cmd(self, cmd):
+        """
+        Executes the supplied AT command, which must be terminated with
+        a new-line character.
+        Returns msgtype, rspid, rsp, which are 8-bit int, 16-bit int and a
+        bytearray.
+        :param cmd: The new-line terminated AT command to execute.
+        """
+        # Make sure we stay within the 20 byte limit
+        if len(cmd) > 16:
+            # TODO: Split command into multiple packets
+            if self._debug:
+                print("ERROR: Command too long.")
+            raise ValueError('Command too long.')
+
+        # Check out the SPI bus
+        while not self._spi.try_lock():
+            pass
```

Comment:
Do you have to lock across these multiple transactions? Otherwise I'd suggest using a context manager to manage the transaction using SPIDevice.

```python
        with self._spi as spi:
            spi.write(self._buf_tx, end=len(cmd) + 4)
```

Reply:
ditto